hilti/src/rt/types/stream.cc

```diff
@@ -156,209 +269,124 @@ void View::copyRaw(Byte* dst) const {
-                // Just set the new head to empty chunk. Note that we need to
-                // keep the current chain object so that iterators don't
-                // become invalid.
-                auto chain = _content;
-                chain->head = chain->tail = std::shared_ptr<Chunk>(new Chunk(i.offset(), {}, 0));
-                return;
-            }
-
-            continue;
-        }
-
-        if ( c->offset() <= i.offset() && i.offset() < c->offset() + c->size() ) {
-            c->trim(i.offset());
-            break;
-        }
-    }
-}
-
-void Stream::freeze() {
-    _frozen = true;
-    for ( auto c = head(); c; c = c->next().get() )
-        c->freeze();
-}
-
-void Stream::unfreeze() {
-    _frozen = false;
-    for ( auto c = head(); c; c = c->next().get() )
-        c->unfreeze();
-}
-
-Stream::Content Stream::deepCopyContent() const {
-    std::shared_ptr<Chunk> head;
-    std::shared_ptr<Chunk> tail;
-
-    for ( auto ch = _content->head; ch; ch = ch->next() ) {
-        auto nch = std::make_shared<Chunk>(*ch);
-        if ( tail )
-            tail->setNext(nch);
-
-        if ( ! head )
-            head = nch;
-
-        tail = nch;
-    }
+Size View::size() const {
+    // Because our end offset may point beyond what's currently
+    // available, we need to take the actual end in account to return
+    // the number of actually available bytes.
 
-    return std::make_shared<stream::detail::Chain>(std::move(head), std::move(tail));
-}
+    if ( ! _begin.chain() )
+        return 0;
 
-Size View::size() const {
-    if ( end().offset() <= _begin.offset() )
+    auto tail = _begin.chain()->tail();
+    if ( ! tail )
         return 0;
 
-    // Not so great: Because our end offset may point beyond what's currently
-    // available, we actually need to iterate through and count.
-    //
-    // TODO(robin): We can build a better loop though.
-    Size s = 0;
-    auto x = end();
-    auto end_ = detail::UnsafeConstIterator(end());
-    for ( auto i = detail::UnsafeConstIterator(_begin); i != end_; ++i )
-        s++;
+    if ( _begin.offset() > tail->endOffset() )
+        return 0;
 
-    return s;
+    if ( ! _end || _end->offset() >= tail->endOffset() )
+        return tail->endOffset() - _begin.offset();
+    else
+        return _end->offset() > _begin.offset() ? (_end->offset() - _begin.offset()).Ref() : 0;
 }
 
-std::string Stream::data() const {
-    std::string s;
-    s.reserve(size());
-
-    for ( auto i = unsafeBegin(); i != unsafeEnd(); ++i )
-        s += static_cast<char>(*i);
-
-    return s;
-}
+std::string Stream::data() const { return view().data(); }
 
 std::string stream::View::data() const {
     std::string s;
     s.reserve(size());
 
-    for ( auto i = unsafeBegin(); i != unsafeEnd(); ++i )
-        s += static_cast<char>(*i);
+    for ( auto block = firstBlock(); block; block = nextBlock(block) )
+        s += std::string(reinterpret_cast<const char*>(block->start), block->size);
```

Comment:
We need to do a `reinterpret_cast` here since `Byte` is a `uint8` (aka a `unsigned char`) while a `std::string` wants to be constructed from a signed `char`. This seems to do what we want, but looks like tech debt to me.

I think ideally this function wouldn't return a `std::string` but a `Bytes` so that we could directly append a range of `Byte` to it (would need a `void Bytes::append(Byte*, size_t)` and probably also a way to `reserve` capacity in a `Bytes`). Such a change percolates a little further than this small fix, but it might be worth it.

Reply:
returns a `bytes` is good, but also a larger change. Will leave as is for now, but will create a ticket for it.